core/variable_map.js

```diff
@@ -0,0 +1,222 @@
+  }
+  if (newVariable){ // see if I can get rid of newVariable dependency
+    newVariableIndex = variableList.indexOf(newVariable);
+  }
+
+  if (variableIndex == -1 && newVariableIndex == -1) {
+    this.createVariable(newName, '');
+    console.log('Tried to rename an non-existent variable.');
+  }
+  else if (variableIndex == newVariableIndex ||
+      variableIndex != -1 && newVariableIndex == -1) {
+    // Only changing case, or renaming to a completely novel name.
+    this.variableMap_[type][variableIndex].name = newName;
+  } else if (variableIndex != -1 && newVariableIndex != -1) {
+    // Renaming one existing variable to another existing variable.
+    // The case might have changed, so we update the destination ID.
+    this.variableMap_[type][newVariableIndex].name = newName;
+    this.variableMap_[type].splice(variableIndex, 1);
+  }
+};
+
+/**
+ * Create a variable with a given name, optional type, and optional id.
+ * @param {!string} name The name of the variable. This must be unique across
+ *     variables and procedures.
+ * @param {?string} opt_type The type of the variable like 'int' or 'string'.
+ *     Does not need to be unique. Field_variable can filter variables based on
+ *     their type. This will default to '' which is a specific type.
+ * @param {?string} opt_id The unique id of the variable. This will default to
+ *     a UUID.
+ */
+Blockly.VariableMap.prototype.createVariable = function(name, opt_type, opt_id) {
+  var variable = this.getVariable(name);
+  if (variable) {
+    if (opt_type && variable.type != opt_type) {
+      throw Error('Variable "' + name + '" is already in use and its type is "'
+                  + variable.type + '" which conflicts with the passed in ' +
+                  'type, "' + opt_type + '".');
+    }
+    if (opt_id && variable.getId() != opt_id) {
+      throw Error('Variable "' + name + '" is already in use and its id is "'
+                  + variable.getId() + '" which conflicts with the passed in ' +
+                  'id, "' + opt_id + '".');
+    }
+    return;
+  }
+  if (opt_id && this.getVariableById(opt_id)) {
+    throw Error('Variable id, "' + opt_id + '", is already in use.');
+  }
+  opt_id = opt_id || Blockly.utils.genUid();
+  opt_type = opt_type || '';
+
+  variable = new Blockly.VariableModel(name, opt_type, opt_id);
+  // If opt_type is not a key, create a new list.
+  if (!this.variableMap_[opt_type]) {
+    this.variableMap_[opt_type] = [variable];
+  }
+  // Else append the variable to the preexisting list.
+  else {
+    this.variableMap_[opt_type].push(variable);
+  }
+};
+
+/**
+ * Delete a variable.
+ * @param {Blockly.VariableModel} variable Variable to delete.
+ */
+Blockly.VariableMap.prototype.deleteVariable = function(variable) {
+  var type = variable.type;
+  for (var i = 0, tempVar; tempVar = this.variableMap_[type][i]; i++) {
+    if (Blockly.Names.equals(tempVar.name, variable.name)) {
+      delete this.variableMap_[type][i];
```

Comment:
I think you don't need to call delete--it should get handled by gc once it's removed from the list, and variableModel doesn't have any special disposal code.